force_calibration: refactor the docs (#466)
lumicks/pylake: docs/tutorial/force_calibration.rst

Force Calibration
=================

.. only:: html

    :nbexport:`Download this page as a Jupyter notebook <self>`

Force calibration refers to computing the conversion factors needed to convert the raw voltages recorded by the position sensitive detectors to actual forces and displacements.
The values we wish to calculate are:

> - Trap stiffness :math:`\kappa`, which tells us how stiff the trap is.

Author: Yeah, good call, I'd say it's a little stronger than just related to though.

Perhaps the following?
```
Trap stiffness :math:`\kappa`, which reflects how strongly a bead is held by a trap
```

- Force response :math:`R_d`, which converts from voltage to force.
- Distance response :math:`R_f`, which converts from voltage to distance.

> This tutorial will focus on performing force calibration with Pylake.

Author: Yes, good point. Added.

For the theoretical background on the physics of force calibration, please refer to :doc:`theory section on force calibration</theory/force_calibration/force_calibration>`.

> Undoing the previous calibration

Author: I have moved part of it up, and included your description.

I did add a little warning box in this section though to at least alert users that there could be an issue there when `f_c` approaches `f_diode`.

--------------------------------

This tutorial covers calibration using Pylake.
Our starting point will be data acquired with Bluelake, which is already calibrated.
Therefore, the first step is to undo the calibration applied by Bluelake to get the raw voltage signals.
Let's load the dataset::

    f = lk.File("passive_calibration.h5")

> We grab the chunk of data that was used for calibration last time::

Author: I don't think this is correct actually. It contains only one calibration section. The zeroth is always the calibration that was active prior to the slice (so that the voltages can be recovered). There is only one actual calibration here.

Still, I think making this more explicit like you have is good. So how about the following?

```
The force timeline in this file contains a section used for calibration.
We can see calibrations relevant to this file by inspecting the `calibration` attribute for the entire force :class:`~lumicks.pylake.channel.Slice`.
This is a list of dictionaries containing all of the relevant calibration parameters::

    print(len(f.force1x.calibration))
    print(f.force1x.calibration[1])

The first calibration always refers to the calibration that was active at the start of the slice (usually from a calibration that occurred before the start of the slice).
Our goal here is to recalibrate the data in the calibration section (corresponding to `calibrations[1]`).
First we grab the chunk of data that was used for that calibration in Bluelake::
```

Downsampling the spectrum is required to fulfill some of the assumptions in the fitting procedure, but it comes at the cost of spectral resolution.
One must be careful that the shape of the power spectrum is still sufficiently preserved.
If the corner frequency is very low then downsampling too much can lead to biases in the calibration parameters.
In such cases, it is better to either measure a longer interval to increase the spectral resolution or reduce the number of points (`num_points_per_block`) used for blocking.

When calculating the power spectrum, the amount of downsampling (`num_points_per_block`) and the range over which to compute the spectrum (`fit_range`) can be controlled.
One can also exclude specific frequency ranges from the spectrum (`excluded_ranges`) which can be useful if there are noise peaks in the spectrum.
Let's see which ranges were excluded in our Bluelake calibration::

    force_slice.calibration[0]

.. image:: figures/force_calibration/bl_dictionary.png

Here, they are listed as `Exclusion range 0 (min.) (Hz)`, `Exclusion range 0 (max.) (Hz)` etc.
To reproduce the result obtained with Bluelake, these should be excluded from the power spectrum::

>     power_spectrum = lk.calculate_power_spectrum(volts.data, sample_rate=volts.sample_rate, fit_range=(1e2, 23e3), num_points_per_block=2000, excluded_ranges=([19348, 19668], [24308, 24548]))

Author: I was also wondering about that actually. I will fix it up.

    power_spectrum.plot(marker=".")

.. image:: figures/force_calibration/power_spectrum_excluded_ranges.png

Note that exclusion ranges are excluded _prior_ to downsampling.
Considering that a noise peak may be very narrow, it is beneficial to lower the number of points per block temporarily to find the exact exclusion range.
After determination of this exclusion range, the number of points per block can be increased again.

Passive calibration
-------------------

In literature, passive calibration is often referred to as thermal calibration.

> It involves fitting a physical model to the power spectrum obtained in the previous step.

Author: Yeah, I kept it generic on purpose, since I wanted to keep the terminology to a minimum. It's also not just a Lorentzian. It's a Lorentzian times a filtering function in the simplest case. I'd rather rephrase it a bit in the fast sensor section to not mention the Lorentzian, but refer to it as the simple model. What do you think?

> It involves fitting a physical model to the power spectrum obtained in the previous step.

Reviewer: maybe could work. You use "Lorentzian" a lot in the theory section too. So either replacing the term everywhere or defining it somewhere I'd be ok with

> It involves fitting a physical model to the power spectrum obtained in the previous step.

Author: I've explicitly adding a sentence defining it in the theory section. I think using terminology in the section where people are interested in theory is more acceptable, no?

This physical model relies on a number of parameters that have to be specified in order to get the correct calibration factors.

The most important of these is the bead diameter (in microns).
Let's use the bead diameter found in the calibration performed in Bluelake.

You can optionally also provide a viscosity (in Pa/s) and temperature (in degrees Celsius).
To find the viscosity of water at a particular temperature, Pylake uses :func:`~lumicks.pylake.viscosity_of_water` which implements the model presented in :cite:`huber2009new`.
When omitted, this function will automatically be used to look up the viscosity of water for that particular temperature

> Note that for experiments that use a different medium than water, the viscosity at the experimental temperature should explicitly be provided.

Author: Yes.

    It can optionally be disabled by passing `bias_correction=False` to :func:`~lumicks.pylake.fit_power_spectrum` or :func:`~lumicks.pylake.calibrate_force`.

We can plot the calibration by calling::

    calibration.plot()

.. image:: figures/force_calibration/fitted_spectrum.png

While the simple theory can suffice for small beads, it is usually a good idea to use the more realistic hydrodynamically correct model.
This model takes into account hydrodynamic and inertial effects (which scale with the size of the bead) leading to more accurate estimates.
As such, it requires a few extra parameters: the density of the sample and bead::

    force_model = lk.PassiveCalibrationModel(bead_diameter, hydrodynamically_correct=True, rho_sample=999, rho_bead=1060.0)

Note that when `rho_sample` and `rho_bead` are omitted, values for water and polystyrene are used for the sample and bead density respectively.

.. note::

    One thing to note is that when using the hydrodynamically correct model, the equation for the drag does include the viscosity and bead diameter.

>     However, they now appear in a term which already amounts to a small correction therefore the impact of any errors in these is reduced :cite:`tolic2006calibration`.

Author: Yes. I think we can actually omit this statement. It is not really relevant for the tutorial, and more for the theory section.

Calibration near the surface
----------------------------

So far, we have only considered experiments performed deep in bulk.
In reality, proximity of the flowcell surface to the bead leads to an increase in the drag force on the bead.
Such surface effects can be taken into account by specifying a distance to the surface (in microns)::

    force_model = lk.PassiveCalibrationModel(bead_diameter, hydrodynamically_correct=True, rho_sample=999, rho_bead=1060.0, distance_to_surface=5)

As we approach the surface, the drag effect becomes stronger.
The hydrodynamically correct model is only valid up to a certain distance to the surface.
Moving closer, the frequency dependent effects become smaller than the overall drag effect and better models to approximate the local drag exist.
We can use this model by setting `hydrodynamically_correct` to `False`, while still providing a distance to the surface::

    force_model = lk.PassiveCalibrationModel(bead_diameter, hydrodynamically_correct=False, distance_to_surface=5)

To summarize, the workflow can be visualized as follows:

> .. image:: figures/force_calibration/surface_calibration_workflow.png

Reviewer 2: I think a visual representation of the of the bead/surface would be good to explain which method to use in which situation, something like this
<img width="839" alt="Screenshot 2023-02-28 at 12 06 45" src="https://user-images.githubusercontent.com/17353250/221836339-bdfc823d-366c-4bd3-a697-c01fb1ef49ee.png">
We could also use it in the bluelake docs.

Author: ![hydro-guide](https://user-images.githubusercontent.com/19836026/222182776-6d4e4bb7-63b7-4e2a-96bc-bdcb7434d951.jpg)
Courtesy of Thana :)

.. note::

    Note that the drag coefficient that `pylake` returns always corresponds to the drag coefficient extrapolated back to its bulk value.

Axial Force
-----------

No hydrodynamically correct model is available for axial calibration.
However, models do exist for the dependence of the drag force on the distance to the surface.
Axial force calibration can be performed by specifying `axial=True`::

    force_model = lk.PassiveCalibrationModel(bead_diameter, distance_to_surface=5, axial=True)

Active calibration
------------------

Active calibration has a few benefits.
When performing passive calibration, we base our calculations on a theoretical drag coefficient.
This theoretical drag coefficient depends on parameters that are only known with limited precision:

- The diameter of the bead :math:`d` in microns.
- The dynamic viscosity :math:`\eta` in Pascal seconds.
- The distance to the surface :math:`h` in microns.

> The viscosity in turn depends strongly on the local temperature around the bead, which is typically poorly known.
> In active calibration, an extra peak with a known amplitude is introduced into the power spectrum which allows the trap to be calibrated without prior knowledge of the drag coefficient.

Reviewer: ```suggestion
The viscosity in turn depends strongly on the local temperature around the bead, which is typically poorly known.

In active calibration, we oscillate the stage with a known frequency and amplitude. This introduces an extra peak in the power spectrum which allows the trap to be calibrated without the assumptions of the theoretical drag coefficient.
```

Author: 🙌

Using `pylake`, the procedure to use active calibration is not very different from passive calibration.
However, it does require some additional data channels as inputs.
In the next section, the aim is to calibrate the x-axis of trap 1.
We will consider that the nanostage was used as driving input.
Let's analyze some active calibration data acquired near a surface.
To do this, load a new file::

    f = lk.File("near_surface_active_calibration.h5")
    volts = decalibrate(f.force1x)
    bead_diameter = f.force1x.calibration[0]["Bead diameter (um)"]
    # Calibration performed at 1.04 * bead_diameter
    distance_to_surface = 1.04 * bead_diameter

> Extract the nanostage data::

Reviewer: ```suggestion
First we need to extract the nanostage data which is used to determine the driving amplitude and frequency::
```

    driving_data = f["Nanostage position"]["X"]

> For an active calibration, we should see an oscillation here.

Reviewer: ```suggestion
For data acquired with active calibration in Bluelake, this will be a sinusoidal oscillation.
```

If there are unexplained issues with the calibration, it is a good idea to plot the driving signal and verify that the motion looks like a clean sinusoid::

    driving_data.plot()
    plt.xlim(0, 0.1)

.. image:: figures/force_calibration/nanostage_position.png

Instead of using the :class:`~lumicks.pylake.PassiveCalibrationModel` presented in the previous section, we now use the :class:`~lumicks.pylake.ActiveCalibrationModel`.
We also need to provide the sample rate at which the data was acquired, and a rough guess for the driving frequency.
`pylake` will find an accurate estimate of the driving frequency based on this initial estimate (provided that it is close enough)::

    active_model = lk.ActiveCalibrationModel(driving_data.data, volts.data, driving_data.sample_rate, bead_diameter, driving_frequency_guess=38, distance_to_surface=distance_to_surface)

> To check the determined frequency, we can look at the determined driving frequency::

Reviewer: ```suggestion
We can check the determined driving frequency with::
```

    >>> active_model.driving_frequency
    38.15193077664577

Let's have a look to see if this peak indeed appears in our power spectrum.
To see it clearly, we reduce the blocking amount and show the spectrum all the way up to a frequency of 10 Hz::

    show_peak = lk.calculate_power_spectrum(volts.data, sample_rate=78125, num_points_per_block=5, fit_range=(10, 23000))
    show_peak.plot()

.. image:: figures/force_calibration/calibration_peak.png

The driving peak is clearly visible in the spectrum.

> Let's calculate the power spectrum we'll use for fitting.
> It is important to _not_ include the driving peak when doing this (the default will only go up to 100 Hz)::

Reviewer: ```suggestion
Next let's calculate the power spectrum we'll use for fitting.
It is important to *not* include the driving peak when doing this (the default will only go up to 100 Hz)::
```

    power_spectrum = lk.calculate_power_spectrum(volts.data, sample_rate=78125)

We can now use this to fit our data::

    calibration = lk.fit_power_spectrum(power_spectrum, active_model)
    calibration

.. image:: figures/force_calibration/calibration_item_active.png

Analogous to the passive calibration procedure, we can specify `hydrodynamically_correct=True` if we wish to use the hydrodynamically correct theory here.
Especially for bigger beads this is highly recommended (more on this later).

More convenient calibration

> ---------------------------

Reviewer: I don't know there's a good way to do this, but I feel this section could be almost at the top. It'd be nice right away to show users that they don't *need* to go through the hassle of setting up the models manually. 

Perhaps move the first half of this section toward the top as a TLDR-ish part and then take the 2nd half here comparing different models as a "Comparison" section here still?

Author: I felt the same way when writing this but held off on moving it because I didn't want to fracture the introduction and use of the convenience function. I do agree though, and will attempt to move it up.

> For convenience, `pylake` offers :func:`~lumicks.pylake.calibrate_force()` which performs the entire calibration procedure in a single function call::

Reviewer: sometimes you use "Pylake" and sometimes `pylake`; I guess we should make a decision on what we want to use and be consistent throughout the docs (at a later time) but we can start with this tutorial

Author: Good point. I doubt with every new instance of Pylake in the docs. I'll make it consistent.

    >>> print(fit["kappa"].value)
    0.11662183772410809

And compare this to the passive calibration result::

    >>> fit = lk.calibrate_force(**shared_parameters, active_calibration=False, distance_to_surface=distance_to_surface)
    >>> print(fit["kappa"].value)
    0.11763849764570819

These values are quite close.
However, if we do not provide the height above the surface, we can see that the passive calibration result suffers much more than the active calibration result (as passive calibration fully relies on a drag coefficient calculated from the physical input parameters)::

    >>> print(lk.calibrate_force(**shared_parameters, active_calibration=False)["kappa"].value)
    >>> print(lk.calibrate_force(**shared_parameters, active_calibration=True)["kappa"].value)
    0.08616565751377737
    0.11662183772410809

.. note::

    When hydrodynamic correction is enabled, the height parameter impacts both passive and active calibration.
    The reason for this is that when using the more accurate theory, the expected shape of the power spectrum depends on the distance to the surface.

>     For more information on this see the :doc:`theory section on force calibration</theory/force_calibration/force_calibration>` section.

Reviewer: this seems a bit strange. In the first sentence you talk about hydrodynamic correction, but in the 2nd distance to surface; the connection isn't really clear

Author: Yes. I agree. I think inverting the argument may be better here. How about:
```
    When fitting with the hydrodynamically correct model, the `distance_to_surface` parameter impacts the expected shape of the power spectrum.
    Consequently, when this model is selected, this parameter affects both passive and active calibration.
    For more information on this see the :doc:`theory section on force calibration</theory/force_calibration/force_calibration>` section.
```

Fast Sensors
------------

> By default, the models used always incorporate a filtering effect.
> This filtering effect occurs because a fraction of the light is not measured instantaneously by most detectors.
> Fast detectors have the ability to respond much faster to incoming light resulting in no visible filtering effect in the frequency range we are fitting.
> This means that for a fast detector, we do not need to include a filtering effect in our model.

Reviewer: If you go with my suggestion for adding an intro section, we don't need the explanation of the filtering effect here.
```suggestion
Fast detectors have the ability to respond much faster to incoming light resulting in no visible filtering effect in the frequency range we are fitting.
This means that for a fast detector, we do not need to include a filtering effect in our model.
```
We might also want to explicitly state that this depends on the particular hardware in the C-Trap (it's implied here, but let's make it super clear).

We can omit this effect by passing `fast_sensor=True` to the calibration models or to :func:`~lumicks.pylake.calibrate_force()`.

> Note however, that this makes using the hydrodynamically correct model critical, as the Lorentzian doesn't actually capture the data very well.

Reviewer: I think this situation needs a bit more explanation; I don't quite get why hydro is more important and if it's dependent on active vs passive

    fit.plot()
    plt.title(f"Lorentzian + Slow (kappa={fit['kappa'].value:.2f})")
    plt.subplot(1, 3, 2)
    fit = lk.calibrate_force(**shared_parameters, hydrodynamically_correct=False, fast_sensor=True, num_points_per_block=200)
    fit.plot()
    plt.title(f"Lorentzian + Fast (kappa={fit['kappa'].value:.2f})")
    plt.subplot(1, 3, 3)
    fit = lk.calibrate_force(**shared_parameters, hydrodynamically_correct=True, fast_sensor=True, num_points_per_block=200)
    fit.plot()
    plt.title(f"Hydro + Fast (kappa={fit['kappa'].value:.2f})")
    plt.tight_layout()

.. image:: figures/force_calibration/fast_sensors.png

Note how the fit with the Lorentzian model with the filtering effect enabled seems to fit the data quite well.
However, the apparent quality of a fit can be deceiving.
When the `fast_sensor` flag is enabled, it can be seen that the Lorentzian model doesn't actually describe the data.
Switching to the hydrodynamically correct model results in a superior fit to the power spectrum.

In the case of the Lorentzian with a detector filtering effect, the fitting procedure has used the parameters of the filter to fit some of the high frequency attenuation.

> With the diode model disabled, we obtain a very biased fit.

Reviewer: here are you using "Lorentzian" to refer to passive calibration? If so, I think we should be consistent and just use passive calibration throughout

Author: No, I use Lorentzian to mean the simple model. I think we should actual hold off on calling it a Lorentzian in the tutorial (and save that for the theory section). I've started referring to it as simple model.

I have also updated the description a bit, and added a question. While I normally try to avoid adding questions like this in explanations, I think in this case it is warranted:

```
Note how the power spectral fit with the simple model seems to fit the data quite well as long as we also include the filtering effect.
However, the apparent quality of a fit can be deceiving.
Considering that this dataset was acquired on a fast sensor, we should omit the filtering effect.
When the `fast_sensor` flag is enabled, it can be seen that the simple model doesn't actually describe the data.
Switching to the hydrodynamically correct model results in a superior fit to the power spectrum.

So what is happening here? Why did the first fit look good?
When we fit the power spectrum with the simple model and include the filtering effect, the fitting procedure uses the parameters that characterize the filter to fit some of the high frequency attenuation.
With the filtering effect disabled, we obtain a very biased fit because the model fails to fit the data.
```

If we compare the different fits, we can see that the Lorentzian model with filtering effect (`fast_sensor=False`) gives similar stiffness estimates as the hydrodynamically correct model without the filtering.
While this is true for this particular dataset, no general statement can be made about the bias caused by fitting a Lorentzian rather than the hydrodynamically correct power spectrum.
If low bias is desired, one should always use the hydrodynamically correct model when possible.
On regular sensors, it is best to fit the hydrodynamically correct model with the filtering effect enabled.

High corner frequencies
-----------------------

In specific situations, the filtering effect of the position sensitive detector can cause issues when calibrating.
The power spectrum of the bead in the optical trap is characterized by a diffusion constant and a corner frequency.

> The filtering effect is characterized by a constant that reflects the fraction of light that is transmitted instantaneously and a corner frequency (referred to as the diode frequency to be able to differentiate).

Author: I like it. I'd prefer actually having part of the section you propose here at the top but with a few changes.
The actual models should not be explained in the tutorial. Those are in the theory section. Therefore I would like to suggest the following amendment (basically refer them to the theory section for the models and the tut for their use):

```
To calibrate the optical traps we fit a power spectrum of the voltage measured by a position sensitive diode (PSD) to a theoretical model. The measured voltage (and thus the shape of the power spectrum) is mainly determined by

1. The Brownian motion of the bead within the trap.
2. The instantaneous response of the PSD to the incident light.

The contribution to the power spectrum by the bead in the optical trap is characterized by a diffusion constant and a corner frequency.
The second contribution refers to a filtering effect where the PSD becomes less sensitive to changes in signal at high frequencies.
This filtering effect is characterized by a constant that reflects the fraction of light that is transmitted instantaneously and a corner frequency (this is referred to as the diode frequency to be able to differentiate).
The various theoretical models that can be used to fit these data are described in the :doc:`theory section on force calibration</theory/force_calibration/force_calibration>`, while their use is described below.

```

.. image:: figures/force_calibration/diode_filtering.png
  :nbattach:

The corner frequency of the physical spectrum can be found in the results as `f_c` and depends on the laser power and bead size (smaller beads resulting in higher corner frequencies) .
The corner frequency of the filtering effect can be found in the results as `f_diode` (which stands for diode frequency) and depends on the incident intensity :cite:`berg2003unintended`.
When these two frequencies get close, they cannot be determined reliably anymore.
The reason for this is that the effect of one can be compensated by the other.
When working with small beads or at high laser powers, it is important to verify that the corner frequency `f_c` does not approach the frequency of the filtering effect `f_diode`.

Sometimes, the filtering effect has been characterized independently.

> In that case, the arguments `fixed_diode` and `fixed_alpha` can be passed to :func:`~lumicks.pylake.calibrate_force()` to fix these parameters to their predetermined values.

Author: I've added a warning box at the top regarding the `f_c`, `f_diode` thing.

- Distance response :math:`R_f`, the proportionality constant between voltage and distance.

This tutorial will focus on performing force calibration with Pylake.
Note that force calibration in Bluelake is actually performed by Pylake.
Therefore, the results obtained with Bluelake can be reproduced exactly when using the same data and parameters.

To calibrate the optical traps we fit a power spectrum of the voltage measured by a position sensitive diode (PSD) to a theoretical model. The measured voltage (and thus the shape of the power spectrum) is mainly determined by

1. The Brownian motion of the bead within the trap.
2. The instantaneous response of the PSD to the incident light.

The contribution to the power spectrum by the bead in the optical trap is characterized by a diffusion constant and a corner frequency.
The second contribution refers to a filtering effect where the PSD becomes less sensitive to changes in signal at high frequencies.
This filtering effect is characterized by a constant that reflects the fraction of light that is transmitted instantaneously and a corner frequency (this is referred to as the diode frequency to be able to differentiate).

> The various theoretical models that can be used to fit these data are described in the :doc:`theory section on force calibration</theory/force_calibration/force_calibration>`, while their use is described below.

Author: Added.

    A warning sign for this is when the corner frequency `f_c` approaches or goes beyond the diode frequency `f_diode`.
    For more information see the section on `High corner frequencies`_.

Undoing the previous calibration
--------------------------------

Our starting point will be data acquired with Bluelake, which is already calibrated.
Therefore, the first step is to undo the calibration applied by Bluelake to get the raw voltage signals.
Let's load the dataset::

    f = lk.File("passive_calibration.h5")

The force timeline in this file contains a section used for calibration.
We can see calibrations relevant to this file by inspecting the `calibration` attribute for the entire force :class:`~lumicks.pylake.channel.Slice`.
This is a list of dictionaries containing all of the relevant calibration parameters::

    print(len(f.force1x.calibration))
    print(f.force1x.calibration[1])

The first calibration always refers to the calibration that was active at the start of the slice (usually from a calibration that occurred before the start of the slice).
Our goal here is to recalibrate the data in the calibration section (corresponding to `calibrations[1]`).

> First we grab the chunk of data that was used for that calibration in Bluelake::

Reviewer: yeah I still find this confusing. I think it could help to use the terms "calibration measurement" and "calibration applied to ....", part of the confusion here I think is just re-used terminology so being more explicit in this case I think is helpful.

But what really made it clear to me was plotting things out. could add a code block to make a figure like this:

![image](https://user-images.githubusercontent.com/61475504/222148375-f081161d-f2a5-4fc2-9314-ede722fff472.png)

the code I used is as follows (but I dunno if it's still too complicated to put in the tutorial, but I don't see a significantly better way to clean it up)

```
force = f.force1x

# final timestamp in data relative to first timestamp, in seconds
stop = (force.timestamps[-1] - force.timestamps[0]) * 1e-9

# measurement timestamp ranges for each calibration relative to first timestamp, in seconds
calibrations = [
    (np.array([cal["Start time (ns)"], cal["Stop time (ns)"]]) - force.timestamps[0]) * 1e-9
    for cal in force.calibration
]

plt.figure()
force.plot()
plt.axvspan(*calibrations[0], alpha=0.5, fc="tab:blue", label="calibration[0] measurement")
plt.axvspan(calibrations[0][1], calibrations[1][0], alpha=0.2, fc="tab:blue", label="calibration[0] applied")
plt.axvspan(*calibrations[1], alpha=0.5, fc="tab:orange", label="calibration[1] measurement")
plt.axvspan(calibrations[1][1], stop, alpha=0.2, fc="tab:orange", label="calibration[0] applied")

plt.legend()
plt.margins(x=0)
# plt.xlim(0, stop)  # uncomment this line to zoom in to data
    
plt.tight_layout()
plt.show()
```

Author: While I like the idea of having this in the docs, I fear it'll distract from the force calibration part of the docs, because people might get hung up on the complexity.

Instead I made the following image, which I propose we use instead (along with some further minor improvements on the wording, differentiating calibration from measurement).

![calibrations](https://user-images.githubusercontent.com/19836026/222209334-c74e657a-c9f8-4e4c-8eaf-32b952ddceb6.png)

        return (force_slice - offset) / response

    volts = decalibrate(force_slice)

Performing the calibration
--------------------------

For convenience, Pylake offers :func:`~lumicks.pylake.calibrate_force()` which performs an entire calibration procedure in a single function call::

    calibration = lk.calibrate_force(
        volts.data,
        force_slice.calibration[0]["Bead diameter (um)"],  # Use the value from the last calibration
        temperature=25,
        sample_rate=volts.sample_rate
    )

This function allows us to quickly try various calibration settings without having to set up all the intermediate models and spectra explicitly.
Note that most parameters have to be provided as keyworded arguments to prevent errors.
This method returns a :class:`~lumicks.pylake.force_calibration.power_spectrum_calibration.CalibrationResults`, which can be plotted using :func:`~lumicks.pylake.force_calibration.power_spectrum_calibration.CalibrationResults.plot()`.

The rest of this tutorial illustrates the various steps involved when performing such a calibration.

Obtaining the power spectrum
----------------------------

> Next, we need the power spectrum to fit. To compute a power spectrum from our data we can invoke :func:`~lumicks.pylake.calculate_power_spectrum()`::

Reviewer: considering the change above, "next" doesn't make sense in this context. We could do something like:
```suggestion
To use the more manual lower-level API, we first need the power spectrum to fit. To compute a power spectrum from our data we can invoke :func:`~lumicks.pylake.calculate_power_spectrum()`::
```
